libp2p/stream/bufferstream.nim

```diff
@@ -93,30 +96,36 @@ type
   AlreadyPipedError* = object of CatchableError
   NotWritableError* = object of CatchableError
 
-proc newAlreadyPipedError*(): ref Exception {.inline.} =
+proc newAlreadyPipedError*(): ref CatchableError {.inline.} =
```

Comment:
exception analysis, if we ever get it, will be more accurate if you return the right type here - generally though, it's better to either use `(ref Xxx)` or `raiseXxx` - the former gives you access to more features, the latter saves on code space for raising, which is already so expensive that you should never actually do it - `inline` is the last thing it should be, just adds bloat

Reply:
Yeah, inline is possibly not the right thing here.

You talking about `(ref MyException(...))` here? Whats the `raiseXxx ` style, I've only come across `newException()`.

Yes, as a general style question, I've always wondered why we do `newXXError` constructor and the return `Exception`? 

Anyways, I didn't make the style its all over our codebase, I'm just trying to be consistent, but thanks for pointing out.

Reply:
`raiseXxx` is in the std lib: https://github.com/nim-lang/Nim/blob/devel/lib/system/arithm.nim#L13 - it should ideally also be annotated with `{.noreturn.}` which helps the compiler in some situations - it saves on code size, generally

> why we do newXXError

no idea, it has literally no advantages - probably a left-over from before object init syntax worked - nobody uses exception tracking in nim, so the feature is kind of broken and unsupported - and if you don't use exception tracking, you don't care what type is returned from a function like this and might as well return the one type that allows you to change what you raise to anything.